### PR TITLE
Fix quarkus-extension-processor race conditions when modules are built in parallel

### LINKED ISSUE #10217: quarkus-extension-processor race conditions when modules are built in parallel
When the Quarkus source tree is built with `mvn -T 1C` or with [mvnd](https://github.com/mvndaemon/mvnd) (which internally uses `-T 1C`), the build intermittently fails either in `quarkus-extension-processor` code or in `io.quarkus.docs.generation.AllConfigGenerator`. This is because quarkus-extension-processor reads and writes files in `<repo-root>/target/asciidoc/generated/config` without any locks or synchronization. When the extension modules are built in parallel, the individual AnnotationProcessor instances overwite eachother's changes. 

Synchronization would be tricky, because the individual AnnotationProcessor instances run in isolated class loaders, so we'd have to synchronize globally on something loaded by the Bootstrap class loader. Locking files would probably lead to a performance bottleneck.

Looking into the details of the data written and read, it seems to be possible to split the files in such a way that each piece is written just once. The ordering of reads after writes should be guaranteed by the module dependency ordering. 

A PR follows.

### PR DESCRIPTION
**Pull request by ppalaga**
…re built in parallel

Fix #10217

### CODE REVIEW

**Comment by ppalaga** [comment]
The CI failure does not seem to be related.

**Comment by ppalaga** [comment]
Any chance to review, @machi1990 ?

**Comment by ppalaga** [comment]
Ad validation of this change: I made sure that this change is not causing any changes in documentation or in bom-descriptor-json/target/extensions.json . Not sure there are other locations that might be impacted.

**Comment by ppalaga** [comment]
Could somebody review pretty please? I would be so nice to get back at the `mvnd` level of productivity.

File: core/processor/src/main/java/io/quarkus/annotation/processor/ExtensionAnnotationProcessor.java [hunk 1]
```diff
@@ -78,6 +78,7 @@
     private final ConfigDocItemScanner configDocItemScanner = new ConfigDocItemScanner();
     private final Set<String> generatedAccessors = new ConcurrentHashMap<String, Boolean>().keySet(Boolean.TRUE);
     private final Set<String> generatedJavaDocs = new ConcurrentHashMap<String, Boolean>().keySet(Boolean.TRUE);
+    private final boolean generateDocs = !Boolean.parseBoolean(System.getProperty("skipDocs", "false"));
```
Comment (gastaldi) [inline core/processor/src/main/java/io/quarkus/annotation/processor/ExtensionAnnotationProcessor.java:81]: Isn't the same as `Boolean.getBoolean("skipDocs")`?

Comment (ppalaga) [reply core/processor/src/main/java/io/quarkus/annotation/processor/ExtensionAnnotationProcessor.java]: Thanks, fixed in 45cd863

**Comment by famod** [comment]
Does this also fix the following?
```
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.8.0-jboss-2:compile (default-compile) on project quarkus-core-deployment: Fatal error compiling: java.lang.IllegalArgumentException: argument "content" is null
```

**Comment by ppalaga** [comment]
> Does this also fix the following?
> 
> ```
> [ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.8.0-jboss-2:compile (default-compile) on project quarkus-core-deployment: Fatal error compiling: java.lang.IllegalArgumentException: argument "content" is null
> ```

Not sure, TBH.

**Comment by ppalaga** [comment]
Added triage/backport in hope that it attracts some reviewers.

**Review by gsmet — CHANGES_REQUESTED** [review]
It doesn't seem to work.

When doing a simple `mvn clean install -DskipTests -DskipITs`, I end up with various warnings, including:
```
[INFO] asciidoctor: INFO: datasource.adoc: line 490: optional include dropped because include file not found: /data/home/gsmet/git/quarkus/target/asciidoc/generated/config/quarkus-datasource.adoc
[INFO] asciidoctor: INFO: datasource.adoc: line 495: optional include dropped because include file not found: /data/home/gsmet/git/quarkus/target/asciidoc/generated/config/quarkus-agroal.adoc
[INFO] asciidoctor: INFO: datasource.adoc: line 583: optional include dropped because include file not found: /data/home/gsmet/git/quarkus/target/asciidoc/generated/config/quarkus-reactive-datasource.adoc
[INFO] asciidoctor: INFO: datasource.adoc: line 587: optional include dropped because include file not found: /data/home/gsmet/git/quarkus/target/asciidoc/generated/config/quarkus-reactive-db2-client.adoc
[INFO] asciidoctor: INFO: datasource.adoc: line 591: optional include dropped because include file not found: /data/home/gsmet/git/quarkus/target/asciidoc/generated/config/quarkus-reactive-mysql-client.adoc
[INFO] asciidoctor: INFO: datasource.adoc: line 595: optional include dropped because include file not found: /data/home/gsmet/git/quarkus/target/asciidoc/generated/config/quarkus-reactive-pg-client.adoc
[INFO] Rendered /data/home/gsmet/git/quarkus/docs/src/main/asciidoc/datasource.adoc
```

**Comment by machi1990** [comment]
Maybe we can use file based locking / synchronization on this part here https://github.com/quarkusio/quarkus/blob/master/core/processor/src/main/java/io/quarkus/annotation/processor/ExtensionAnnotationProcessor.java#L238-L246 without impacting much on how files are generated. 

Yes, we'll incur a small performance cost but that will be better than the race condition.

**Comment by ppalaga** [comment]
> When doing a simple `mvn clean install -DskipTests -DskipITs`, I end up with various warnings, including:
> 
> ```
> [INFO] asciidoctor: INFO: datasource.adoc: line 490: optional include dropped because include file not found: /data/home/gsmet/git/quarkus/target/asciidoc/generated/config/quarkus-datasource.adoc
> [INFO] asciidoctor: INFO: datasource.adoc: line 495: optional include dropped because include file not found: /data/home/gsmet/git/quarkus/target/asciidoc/generated/config/quarkus-agroal.adoc
> [INFO] asciidoctor: INFO: datasource.adoc: line 583: optional include dropped because include file not found: /data/home/gsmet/git/quarkus/target/asciidoc/generated/config/quarkus-reactive-datasource.adoc
> [INFO] asciidoctor: INFO: datasource.adoc: line 587: optional include dropped because include file not found: /data/home/gsmet/git/quarkus/target/asciidoc/generated/config/quarkus-reactive-db2-client.adoc
> [INFO] asciidoctor: INFO: datasource.adoc: line 591: optional include dropped because include file not found: /data/home/gsmet/git/quarkus/target/asciidoc/generated/config/quarkus-reactive-mysql-client.adoc
> [INFO] asciidoctor: INFO: datasource.adoc: line 595: optional include dropped because include file not found: /data/home/gsmet/git/quarkus/target/asciidoc/generated/config/quarkus-reactive-pg-client.adoc
> [INFO] Rendered /data/home/gsmet/git/quarkus/docs/src/main/asciidoc/datasource.adoc
> ```

Thanks for the feedbak @gsmet! I have fixed those in d8c7962. I also made sure that `docs/target/generated-docs` is the same before and after the change.

**Comment by ppalaga** [comment]
> Maybe we can use file based locking / synchronization on this part here https://github.com/quarkusio/quarkus/blob/master/core/processor/src/main/java/io/quarkus/annotation/processor/ExtensionAnnotationProcessor.java#L238-L246 without impacting much on how files are generated.

I vote for considering the current non-locking proposal as long as there is any hope that it can work correctly.

**Comment by ppalaga** [comment]
Let me check what's up with `-Ddocumentation-pdf`

**Comment by ppalaga** [comment]
Hm... I cannot reproduce the error seen in https://github.com/quarkusio/quarkus/pull/10218/checks?check_run_id=866741427#step:8:46462 locally. 7b364b6 is just rebased to see if the error will occur again.

**Comment by ppalaga** [comment]
OK, the CI passed now. Could you please re-review?

**Review by gsmet — APPROVED** [review]

**Comment by gsmet** [comment]
Normal builds look OK now. Let's merge it and see if we detect any further issues.

**Comment by ppalaga** [comment]
Perfect, thanks!

**Pull request merged**